### PR TITLE
core mode: add version()

### PR DESCRIPTION
**Pull request by samouri**
**summary**

Adds `.version` to `#core/mode`. Removes it from `src/internal-version.js`.

### CODE REVIEW

**Review by rcebulko — APPROVED** [review]
I know this used to live in mode, but WDYT about moving it into `#core/constants`? Then again, I suppose the same could be said of all the mode constants, so I'm not sure what makes the most sense. Approving to unblock.

**Comment by samouri** [comment]
> I know this used to live in mode, but WDYT about moving it into `#core/constants`? Then again, I suppose the same could be said of all the mode constants, so I'm not sure what makes the most sense. Approving to unblock.

What is mode but a set of constants :)
I'd be fine with moving _all of mode_ to that.

**Comment by rcebulko** [comment]
> > I know this used to live in mode, but WDYT about moving it into `#core/constants`? Then again, I suppose the same could be said of all the mode constants, so I'm not sure what makes the most sense. Approving to unblock.
> 
> What is mode but a set of constants :)

Things like minified/esm/fortesting etc are really the "mode" in which it was compiled, whereas the RTV is very much a constant. My feeling is that `getMode` got overloaded because it became the place to dump compile-time-ish constants, but really I think this could/should just be a `const INTERNAL_RTV = "...";` in `#core/constants/internal-rtv`, and bring `mode` back to just representing the binary/execution mode

**Comment by samouri** [comment]
> > > I know this used to live in mode, but WDYT about moving it into `#core/constants`? Then again, I suppose the same could be said of all the mode constants, so I'm not sure what makes the most sense. Approving to unblock.
> > 
> > 
> > What is mode but a set of constants :)
> 
> Things like minified/esm/fortesting etc are really the "mode" in which it was compiled, whereas the RTV is very much a constant. My feeling is that `getMode` got overloaded because it became the place to dump compile-time-ish constants, but really I think this could/should just be a `const INTERNAL_RTV = "...";` in `#core/constants/internal-rtv`, and bring `mode` back to just representing the binary/execution mode

I'd push back on the concept of "mode it is compiled on". I don't know of a good definition for what constitutes a mode and what doesn't.  Instead I'd argue the clean separation of constants exists between build constants vs. source constants.

**Comment by rcebulko** [comment]
> I'd push back on the concept of "mode it is compiled in". I don't know of a good definition for what constitutes a mode and what doesn't. Instead I'd argue the clean separation of constants exists between build constants vs. source constants.

I don't feel strongly enough, this is fine by me 👍🏻

**Review by rcebulko — COMMENTED** [review]
I'm curious if it would be sensible to rename this to just `mode.rtv()`? I feel like it's a ubiquitous-enough term that it would be very easily understood

**Comment by samouri** [comment]
> I'm curious if it would be sensible to rename this to just `mode.rtv()`? I feel like it's a ubiquitous-enough term that it would be very easily understood

I'm fine with that.

Separate note: I think we can make the `amp-mode` transform also operate on the namespace import of `mode from #core/mode`

**Comment by rcebulko** [comment]
> > I'm curious if it would be sensible to rename this to just `mode.rtv()`? I feel like it's a ubiquitous-enough term that it would be very easily understood
> 
> I'm fine with that.
> 
> Separate note: DCE doesn't seem to be working. I think we can make the `amp-mode` transform also operate on the namespace import of `import * as mode from #core/mode`

Working for what? DCE should be working for the true/false values; I don't believe it would inline the RTV string if that's what you're expecting.

**Review by jridgewell — APPROVED** [review]

**Comment by samouri** [comment]
> > I'm curious if it would be sensible to rename this to just `mode.rtv()`? I feel like it's a ubiquitous-enough term that it would be very easily understood
> 
> I'm fine with that.
> 
> Separate note: DCE doesn't seem to be working. I think we can make the `amp-mode` transform also operate on the namespace import of `import * as mode from #core/mode`

Actually...this isn't `rtv()`, it is `version()`...updated accordingly. Also will update the babel plugin in the same PR so we don't have a size regression.

File: build-system/babel-plugins/babel-plugin-amp-mode-transformer/index.js
Comment (rcebulko) [inline build-system/babel-plugins/babel-plugin-amp-mode-transformer/index.js:37]: Would the plugin currently handle if say `#core/foobar` had an `import * as mode from './mode';`?

**Review by rcebulko — APPROVED** [review]
This is great! Thanks for doing it
Side note: I've found it easier to shift logic to `core` in one PR and have `mode.js` reference it, then (in theory) eventually update callsites/transforms in a separate PR. If you feel motivated to keep working on these, uses of `getMode(win).test`, and `getMode(win).localDev` could both be updated to the core versions as well; I moved the logic to core, but didn't update all the files because of OWNERS approval 😆

File: build-system/babel-plugins/babel-plugin-amp-mode-transformer/index.js
Comment (samouri) [reply build-system/babel-plugins/babel-plugin-amp-mode-transformer/index.js:37]: nope :/

Comment (rcebulko) [reply build-system/babel-plugins/babel-plugin-amp-mode-transformer/index.js:37]: Could we solve that?

Comment (samouri) [reply build-system/babel-plugins/babel-plugin-amp-mode-transformer/index.js:37]: sure, we could resolve the filename and then check against `src/core/mode.js`, but given that we plan to always use `import * as mode from '#core/mode'` I thought it wouldn't be that important

Comment (rcebulko) [reply build-system/babel-plugins/babel-plugin-amp-mode-transformer/index.js:37]: If it's not much more complexity to do so, I feel like it could be worthwhile (there are also files in `src/` still that import from core via relative path), but the marginal gain is pretty tiny so if it's more than a few minutes to tweak I say 🚢 it

Comment (samouri) [reply build-system/babel-plugins/babel-plugin-amp-mode-transformer/index.js:37]: This PR is blocked on https://github.com/ampproject/amphtml/pull/35477

Comment (rcebulko) [reply build-system/babel-plugins/babel-plugin-amp-mode-transformer/index.js:37]: SGTM.
Doing a find-replace on `%s/from '([^']*/core|\.)/mode';/from '#core/mode';` would probably cover the remaining cases

**Comment by Mstafamhamadmhamad** [comment]
> Hey @jridgewell! These files were changed:
> 
> ```
> src/core/mode/index.js
> src/core/mode/version.js
> ```

**Pull request merged**
